examples/graphql/resolvers.php

```diff
@@ -57,7 +59,17 @@
         GraphQL\publish('inbox', $message);
         // <- Exactly what "inbox" will receive
         return $message;
-    }
+    },
+    'upload' => function () {
+        /** @var array<string, mixed> $file */
+        $file = request()->files[0];
+        $filename = array_get_str($file, 'name');
+        $tmp_name = array_get_str($file, 'tmp_name');
+
+        move_uploaded_file($tmp_name, __DIR__ . '/uploads/' . $filename);
+
+        return ['filename' => $filename];
+    },
```

Comment:
@jordanbrauer can I request you for comments here?

Reply:
Is there any sanitizing done on the file name?
Is there any check to see if the file type matches the extension?
Are there any limits on the file types that can be uploaded?

Reply:
Sure thing! Taking a look, now!

Reply:
@joyously thank you for joining the discussion! No, there is no such checks, just multipart/form-data handling of what are the operations and what is file data, based on [graphql-multipart-request-spec's request-payload](https://github.com/jaydenseric/graphql-multipart-request-spec#request-payload).
Must Siler as a GraphQL server framework handle this or leaving sanitation and validation for user-land at resolvers is ok?

Reply:
If you provide it, it should be secure. If you leave it to user-land, you get a mix, and everyone has to figure it out themselves.
I would at the least prevent directory traversal (like a file name of '../../../etc/passwds' ) and special characters that will cause problems (like null and \n and such).

Reply:
Since it is an example file, I imagine that you would leave this kind of implementation to the user of Siler? At the company I work for, we do a similar thing as in the example for reading multipart form data in GraphQL, except we have our request instance on the context object (but this is just example code anyways) – the tricky part of it is on the client side, so beyond reading the file input from the request on the server, there isn't much else. 

---

@joyously here is my thoughts on your questions – would love to hear what you think too!

> Is there any sanitizing done on the file name?

I think that this would be left up to the user to implement as part of their application. In a real world resolver, the data would typically be passed to some kind of service layer which would handle input sanitization & validation.

> Is there any check to see if the file type matches the extension?
> Are there any limits on the file types that can be uploaded?

This too would be something left to the user to validate. There are many ways it could be done, such as the example from above where an application's service layer would handle this; another way is using schema directives (adding in #267) as a way to add file type validation in GraphQL. You could implement an `@mime` directive which throws an error if the field receives invalid input, and _might_ look something like this...

```graphql
directive @mime(type: [MimeTypeEnum!]!) on FIELD_DEFINITION

enum MimeTypeEnum {
    TextPlain
    MultipartFormData
    ImageGif
    ImageJpg
    # etc. ...
}

type Query {
    uploadImage(file: Upload!): File! @mime(type: [ImageGif ImageJpg])
}
```

We could take a [look at another PHP GraphQL's `Upload` scalar implementation](https://github.com/nuwave/lighthouse/blob/master/src/Schema/Types/Scalars/Upload.php) and see that they are using Laravel's `UploadedFile` class as the determining factor for type validation. Arguably, _Siler_ could provide an `Upload` scalar class implementation, but that is a question for @leocavalcante (perhaps in new MR add custom scalar support, sort of like the directives in #267?).

Reply:
If it is just an example, be sure to put enough comments pointing to what needs to be handled by the user, so it is clear that this is not secure.
But since it returns the file name, it seems like it should clean it up.

Reply:
Thank you so much guys for taking some of your time to share your thoughts! I really appreciate it.
I'll be doing at least the mapping part so the user retrieves the file from the `$args` and implement a default `Upload` type; lefting sanitation and validation to user-land, but being careful on the exemples, probably including sanitation and validation on it.